keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,372 @@
+RestartContainer policy takes precedence over NoRestart policy.
+
+Additionally, if RestartPolicy is 'Never', ResizePolicy should be set to
+NoRestart in order to pass validation.
+
+#### CRI Changes
+
+Kubelet calls UpdateContainerResources CRI API which currently takes
+*runtimeapi.LinuxContainerResources* parameter that works for Docker and Kata,
+but not for Windows. This parameter changes to *runtimeapi.ContainerResources*,
+that is runtime agnostic, and will contain platform-specific information.
+
+### Kubelet and API Server Interaction
+
+When a new Pod is created, Scheduler is responsible for selecting a suitable
+Node that accommodates the Pod.
+
+For a newly created Pod, Spec.Containers[i].ResourcesAllocated must match
+Spec.Containers[i].Resources.Requests. When Kubelet admits a new Pod, values in
+Spec.Containers[i].ResourcesAllocated are used to determine if there is enough
+room to admit the Pod. Kubelet does not set Pod's ResourcesAllocated after
+admitting a new Pod.
+
+When a Pod resize is requested, Kubelet attempts to update the resources
+allocated to the Pod and its Containers. Kubelet first checks if the new
+desired resources can fit the Node allocable resources by computing the sum of
+resources allocated (Pod.Spec.Containers[i].ResourcesAllocated) for all Pods in
+the Node, except the Pod being resized. For the Pod being resized, it adds the
+new desired resources (i.e Spec.Containers[i].Resources.Requests) to the sum.
+* If new desired resources fit, Kubelet accepts the resize by updating
+  Pod.Spec.Containers[i].ResourcesAllocated via pods/resourceallocation
+  subresource, and then proceeds to invoke UpdateContainerResources CRI API
+  to update the Container resource limits. Once all Containers are successfully
+  updated, it updates Pod.Status.ContainerStatuses[i].Resources to reflect the
+  new resource values.
+* If new desired resources don't fit, Kubelet rejects the resize, and no
+  further action is taken.
+  - Kubelet retries the Pod resize at a later time.
+
+Scheduler may, in parallel, assign a new Pod to the Node because it uses cached
+Pods to compute Node allocable values. If this race condition occurs, Kubelet
+resolves it by rejecting that new Pod if the Node has no room after Pod resize.
+
+#### Kubelet Restart Tolerance
+
+If Kubelet were to restart amidst handling a Pod resize, then upon restart, all
+existing Pods, new & departed Pods, and Pods needing resize are handled as follows:
+
+1. All existing Pods (including Pods needing resources resize) are admitted at
+   their current Pod.Spec.Containers[i].ResourcesAllocated values, with Pods
+   sorted by their creation time.
+1. All departed Pods are cleaned up.
+1. Once all existing (and departed) Pods are accounted for, new Pods & Pods
+   needing resize (i.e Pods for which Spec.Containers[i].Resources.Requests
+   and Spec.Containers[i].ResourcesAllocated differ) are sorted by increasing
```

Comment:
I'm not entirely sure how we would implement this given the current structure of the kubelet.  I don't think this level of detail is strictly required for the KEP to move forward.  Can we just say that "Pods requiring a resize are handled after all existing pods have been readmitted"?  I think that is the minimum statement to ensure correctness around restarts.

If it turns out to be feasible to implement sorting of pending updates, we can just add that during the implementation phase...

Reply:
Sure I'll restate it to the minimum. Without prototyping it, I can't say with conviction if ResourceVersions is feasible - though looking at code it seemed possible.

Reply:
Fixed in updated KEP. Resolving issue.